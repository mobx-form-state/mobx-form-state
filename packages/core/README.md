# mobx-form-state/core

[![npm version](https://img.shields.io/npm/v/@mobx-form-state/core.svg?style=flat-square)](https://www.npmjs.org/package/@mobx-form-state/core)
[![npm downloads](https://img.shields.io/npm/dm/@mobx-form-state/core.svg?style=flat-square)](http://npm-stat.com/charts.html?package=mobx-form-state/core)
[![Build Status](https://travis-ci.org/mobx-form-state/core.svg?branch=master)](https://travis-ci.org/@mobx-form-state/core)

## Core package

[Documentation](https://mobx-form-state.netlify.com)
