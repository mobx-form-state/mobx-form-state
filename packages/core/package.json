{
  "name": "@mobx-form-state/core",
  "version": "1.2.0",
  "description": "core package",
  "keywords": [
    "mobx",
    "form"
  ],
  "author": "alexolefirenko <olefirenk@gmail.com>",
  "homepage": "https://mobx-form-state.netlify.com",
  "license": "MIT",
  "main": "./index.js",
  "typings": "./index.d.ts",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/mobx-form-state/mobx-form-state.git"
  },
  "scripts": {
    "copy:files": "ts-node ../../scripts/copy-files",
    "prepublish": "ts-node ../../scripts/copy-package-json"
  },
  "bugs": {
    "url": "https://github.com/mobx-form-state/mobx-form-state/issues"
  },
  "dependencies": {
    "fast-deep-equal": "^3.1.3"
  },
  "peerDependencies": {
    "mobx": "^6.0.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "gitHead": "f2d00255caee1bc020f741adbe1ab1074c465e73"
}
