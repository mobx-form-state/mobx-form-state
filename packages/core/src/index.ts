export * from './form';
export * from './decorators';
export * from './field';
export * from './field-array';
export * from './fields';
export * from './form-control';
export * from './hash-name';
export * from './validate';
export * from './values';
export * from './normalize';
