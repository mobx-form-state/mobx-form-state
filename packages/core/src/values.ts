/* eslint-disable @typescript-eslint/no-explicit-any */
import { runInAction } from 'mobx';

import { FieldArrayType } from './field-array';
import { HashName } from './hash-name';
import { Callback, Maybe } from './types';

export namespace Values {
  export type ChangeEvent<TValue> =
    | TValue
    | {
        target: {
          value?: TValue;
          checked?: TValue;
          options?: { selected: boolean; value: TValue extends any[] ? TValue[number] : TValue }[];
          type?: 'checkbox' | 'select-multiple' | 'text';
        };
      };

  export type SubmitEvent = {
    preventDefault?: Callback;
  };

  export const createEmpty = <MValue>(): MValue => {
    return {} as MValue;
  };

  export const getIn = <TValue, MValue>(
    iterableHash: HashName.HashIterator<keyof MValue>[],
    values: MValue
  ): TValue => {
    return runInAction(
      () =>
        iterableHash.reduce((acc: unknown, item) => {
          if (isValues<MValue>(acc)) {
            switch (item.type) {
              case 'array-item': {
                if ((acc as unknown as []).length <= Number(item.name)) {
                  (acc as unknown as []).push({} as never);
                }

                return acc[item.name];
              }

              case 'object': {
                if (!(item.name in acc)) {
                  acc[item.name] = {} as MValue[keyof MValue];
                }

                return acc[item.name];
              }

              case 'array':
                if (!(item.name in acc)) {
                  acc[item.name] = [] as unknown as MValue[keyof MValue];
                }

                return acc[item.name];
            }
          }

          return acc;
        }, values) as TValue
    );
  };

  export const isValues = <MValue>(x: unknown): x is MValue => typeof x === 'object';

  export type Errors<MValue> = {
    [P in keyof MValue]?: MValue[P] extends Maybe<FieldArrayType<infer FAT>> ? Errors<FAT>[] : string;
  };

  export type FromModel<MValue> = {
    [P in keyof MValue]: MValue[P] extends Maybe<FieldArrayType<infer FAT>> ? FromModel<FAT>[] : MValue[P];
  };

  export const fromEvent = <TValue>(e: ChangeEvent<TValue>): TValue => {
    if (typeof e === 'object' && e !== null && 'target' in e) {
      switch (e.target.type) {
        case 'checkbox':
          return e.target.checked as TValue;
        case 'select-multiple': {
          const value = [];

          if (e.target.options) {
            for (const option of e.target.options) {
              if (option.selected) {
                value.push(option.value);
              }
            }
          }

          return value as unknown as TValue;
        }

        default:
          return e.target.value as TValue;
      }
    }

    return e as unknown as TValue;
  };
}
