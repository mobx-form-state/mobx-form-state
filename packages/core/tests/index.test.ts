/* eslint-disable @typescript-eslint/no-explicit-any */
import { autorun } from 'mobx';
import { Form } from '@mobx-form-state/core';
import { Values } from '@mobx-form-state/core/values';

import { UserFormModel } from './UserForm.model';

describe('simple example', () => {
  test('initialize from model', () => {
    const form = new Form(UserFormModel, {
      name: 'UserForm',
    });

    expect(form.values).toMatchSnapshot();
    expect(form.name).toBe('UserForm');
  });

  test('initialValues should override model default values', () => {
    const form = new Form(UserFormModel, {
      initialValues: {
        login: 'Alex',
        password: 'Password',
        hobbies: [{ hobbyId: 1, hobbyName: 'Dev' }],
        emails: ['Olefirenk@gmail.com'],
      },
    });

    expect(form.values).toMatchSnapshot();
  });

  test('Field: onChange', () => {
    const form = new Form(UserFormModel);

    form.fields.login.onChange({ target: { value: 'test' } });

    expect(form.values.login).toBe('test');
    expect(form.fields.login.value).toBe('test');
  });

  test('FieldArray: validate', async () => {
    const form = new Form(UserFormModel, {
      initialValues: {
        login: '',
        password: '',
        hobbies: [],
      },
    });

    form.fields.hobbies.push({ hobbyId: 0, hobbyName: '' });

    await form.validate();

    expect(form.errors).toEqual({
      hobbies: [{ hobbyId: 'Required', hobbyName: 'Required' }],
      login: 'Required',
      password: 'Required',
    });

    form.fields.hobbies.push({ hobbyId: 1, hobbyName: 'dev' });

    await form.validate();

    expect(form.errors).toEqual({
      hobbies: [{ hobbyId: 'Required', hobbyName: 'Required' }, {}],
      login: 'Required',
      password: 'Required',
    });
  });

  test('Field normalize', () => {
    const form = new Form(UserFormModel);

    form.fields.password.onChange({ target: { value: ' testvalue ' } });

    expect(form.fields.password.value).toBe('testvalue');
    expect(form.values.password).toBe('testvalue');
  });

  test('Form reset', () => {
    const form = new Form(UserFormModel, {
      initialValues: {
        password: 'password',
      },
    });

    expect(form.pristine).toBe(true);

    form.fields.password.onChange('Password');

    expect(form.fields.password.value).toBe('Password');
    expect(form.values.password).toBe('Password');
    expect(form.pristine).toBe(false);

    form.reset();

    expect(form.fields.password.value).toBe('password');
    expect(form.values.password).toBe('password');

    expect(form.pristine).toBe(true);
  });

  test('Form submit', async () => {
    // without submit handlers

    const form = new Form(UserFormModel);

    await form.handleSubmit();

    expect(form.valid).toBe(true);
    expect(form.invalid).toBe(false);

    expect(form.submitted).toBe(false);
    expect(form.submitFailed).toBe(false);
    expect(form.submitting).toBe(false);
  });

  test('Form submit :: success', async () => {
    const onSubmit = jest.fn<Promise<any>, [Values.FromModel<UserFormModel>]>();
    const onSubmitSuccess = jest.fn<void, [any]>();

    const form = new Form(UserFormModel, {
      onSubmit,
      onSubmitSuccess,
      initialValues: {
        login: 'user',
        password: 'password',
      },
    });

    const preventDefault = jest.fn();

    await form.handleSubmit({ preventDefault });

    expect(preventDefault).toBeCalled();
    expect(form.valid).toBe(true);
    expect(form.invalid).toBe(false);
    expect(form.submitted).toBe(true);
    expect(form.submitFailed).toBe(false);
    expect(form.submitting).toBe(false);

    expect(onSubmit).toBeCalledWith(form.values, form);
    expect(onSubmitSuccess).toBeCalled();
  });

  test('Form submit :: success without handlers', async () => {
    const onSubmit = jest.fn<Promise<any>, [Values.FromModel<UserFormModel>]>();

    const form = new Form(UserFormModel, {
      onSubmit,
    });

    await form.handleSubmit();

    expect(onSubmit).toBeCalledWith(form.values, form);
  });

  test('Form submit :: validation errors', async () => {
    const onSubmit = jest.fn<Promise<any>, [Values.FromModel<UserFormModel>]>();
    const onSubmitFail = jest.fn<void, [any]>();

    const form = new Form(UserFormModel, {
      onSubmit,
      onSubmitFail,
      initialValues: {
        login: 'user',
        password: 'password',
      },
    });

    expect(form.pristine).toBe(true);
    expect(form.dirty).toBe(false);
    expect(form.valid).toBe(true);

    form.fields.login.onChange('');
    form.fields.password.onChange('');

    await form.validate();
    await form.handleSubmit();

    expect(form.pristine).toBe(false);
    expect(form.dirty).toBe(true);
    expect(form.valid).toBe(false);
    expect(form.invalid).toBe(true);
    expect(form.submitted).toBe(false);
    expect(form.submitFailed).toBe(true);
    expect(form.submitting).toBe(false);

    expect(onSubmit).not.toBeCalledWith(form);
    expect(onSubmitFail).toBeCalledWith(form.errors, form);

    expect(form.errors.login).toBe('Required');
    expect(form.errors.password).toBe('Required');
  });

  test('Form submit :: validation errors without handlers', async () => {
    const onSubmit = jest.fn<Promise<any>, [Values.FromModel<UserFormModel>]>();

    const form = new Form(UserFormModel, {
      onSubmit,
    });

    form.fields.login.onChange('');
    form.fields.password.onChange('');

    await form.handleSubmit();
  });

  test('Form submit :: submit errors', async () => {
    const onSubmit = (): Promise<any> => Promise.reject({ message: 'Wrong credentials' });
    const onSubmitFail = jest.fn();

    const form = new Form(UserFormModel, {
      onSubmit,
      onSubmitFail,
      initialValues: {
        login: 'wrong',
        password: 'password',
      },
    });

    await form.handleSubmit();

    expect(form.submitted).toBe(false);
    expect(form.submitFailed).toBe(true);
    expect(form.submitting).toBe(false);

    expect(onSubmitFail).toBeCalledWith({ message: 'Wrong credentials' }, form);
  });

  test('Form submit :: submit errors without handlers', async () => {
    const onSubmit = (): Promise<void> => Promise.reject({ message: 'Wrong credentials' });

    const form = new Form(UserFormModel, {
      onSubmit,
    });

    await form.handleSubmit();
  });

  test('Field input', () => {
    const form = new Form(UserFormModel, {
      initialValues: {
        hobbies: [
          {
            hobbyId: 1,
            hobbyName: 'Dev',
          },
        ],
      },
    });

    const loginField = form.fields.login;

    expect(loginField.active).toBe(false);
    expect(loginField.valid).toBe(true);
    expect(loginField.invalid).toBe(false);
    expect(loginField.error).toBe(undefined);
    expect(loginField.touched).toBe(false);
    expect(loginField.visited).toBe(false);
    expect(loginField.dirty).toBe(false);
    expect(loginField.pristine).toBe(true);

    loginField.onFocus();

    expect(form.fields.login.active).toBe(true);
    expect(form.fields.login.visited).toBe(true);

    loginField.onChange({ target: { value: 'new value' } });

    expect(form.fields.login.pristine).toBe(false);
    expect(form.fields.login.dirty).toBe(true);

    expect(loginField.value).toBe('new value');

    loginField.onBlur();
    expect(form.fields.login.touched).toBe(true);
    expect(form.fields.login.active).toBe(false);
  });

  test('Field checkbox', () => {
    const form = new Form(UserFormModel);

    const employed = form.fields.employed;

    employed.onChange({ target: { type: 'checkbox', checked: false } });

    expect(employed.value).toBe(false);
    expect(form.fields.employed.active).toBe(false);
    expect(form.fields.employed.touched).toBe(true);
  });

  test('Field radio', () => {
    const form = new Form(UserFormModel);

    let mustardRadio = form.fields.sause.bind.radio('mustard');

    let ketchupRadio = form.fields.sause.bind.radio('ketchup');

    autorun(() => {
      mustardRadio = { ...form.fields.sause.bind.radio('mustard') };
    });

    autorun(() => {
      ketchupRadio = { ...form.fields.sause.bind.radio('ketchup') };
    });

    mustardRadio.onChange({ target: { type: 'text', value: 'mustard' } });
    expect(mustardRadio.checked).toBe(true);
    expect(ketchupRadio.checked).toBe(false);
    expect(form.fields.sause.value).toBe('mustard');

    expect(form.fields.sause.active).toBe(false);
    expect(form.fields.sause.touched).toBe(true);
  });

  test('Field select-multiple', () => {
    const form = new Form(UserFormModel, {
      initialValues: {
        flavor: [],
      },
    });

    expect(form.fields.flavor.value).toEqual([]);

    form.fields.flavor.onChange({
      target: {
        options: [
          { selected: true, value: 'grapefruit' },
          { selected: false, value: 'lime' },
        ],
        type: 'select-multiple',
      },
    });

    expect(form.fields.flavor.value).toEqual(['grapefruit']);

    form.fields.flavor.onChange({
      target: {
        type: 'select-multiple',
      },
    });

    expect(form.fields.flavor.value).toEqual([]);
  });

  test('FieldArray values should be consitent', () => {
    const form = new Form(UserFormModel, {
      initialValues: {
        emails: ['olefirenk@gmail.com'],
        hobbies: [
          {
            hobbyId: 1,
            hobbyName: 'Dev',
          },
          {
            hobbyId: 2,
            hobbyName: 'Smth',
          },
        ],
      },
    });

    expect(form.values.hobbies).toEqual(form.initialValues.hobbies);
  });

  test('Form generic error', async () => {
    const handleSubmit = jest.fn(({ login }: Values.FromModel<UserFormModel>) => {
      if (login === 'Alex') return true;

      throw new Error('User not exist');
    });

    const form = new Form(UserFormModel, {
      initialValues: {
        login: 'nonExistLogin',
      },
      onSubmit: handleSubmit,
    });

    await form.handleSubmit();

    expect(form.submitFailed).toBeTruthy();
    expect(form.errorMessage).toBe('User not exist');
  });
});
