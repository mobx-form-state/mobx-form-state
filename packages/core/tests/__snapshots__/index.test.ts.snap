// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple example initialValues should override model default values 1`] = `
Object {
  "emails": Array [
    "Olefirenk@gmail.com",
  ],
  "employed": false,
  "flavor": Array [],
  "hobbies": Array [
    Object {
      "createdAt": 2022-05-29T00:00:00.000Z,
      "hobbyId": 1,
      "hobbyName": "Dev",
    },
  ],
  "login": "Alex",
  "password": "Password",
  "sause": "",
}
`;

exports[`simple example initialize from model 1`] = `
Object {
  "emails": Array [
    "olefirenk@gmail.com",
  ],
  "employed": false,
  "flavor": Array [],
  "hobbies": Array [],
  "login": "login",
  "password": "password",
  "sause": "",
}
`;
